File: app/src/main/java/com/softwareoverflow/hiitultimate/database/entity/WorkoutSetEntity.java
package com.softwareoverflow.hiitultimate.database.entity;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.PrimaryKey;

@Entity(tableName = "workoutSet",
        foreignKeys = {
                @ForeignKey(entity = WorkoutEntity.class, // Where the fk maps to
                        parentColumns = {"workoutId"}, // Column name in the entity mapped to
                        childColumns = {"fkWorkoutId"}, // Column name to map to the entity
                        onUpdate = ForeignKey.CASCADE,
                        onDelete = ForeignKey.CASCADE //Delete this when the WorkoutEntity is deleted
                ),
                @ForeignKey(entity = ExerciseTypeEntity.class, // Where the fk maps to
                        parentColumns = {"exerciseTypeId"}, // Column name in the entity mapped to
                        childColumns = {"fkExerciseTypeId"}, // Column name to map to the entity
                        onUpdate = ForeignKey.CASCADE // Updates to the exercise type are reflected in here
                )
        })
public class WorkoutSetEntity {

    @PrimaryKey(autoGenerate = true)
    private long workoutSetId;

    @ColumnInfo(name = "fkWorkoutId", index = true)
    private long workoutId;

    @ColumnInfo(name = "fkExerciseTypeId", index = true)
    private long exerciseTypeId;
    private int workTime, restTime, numReps, recoveryTime;

    public WorkoutSetEntity(int workTime, int restTime, int numReps, int recoveryTime) {
        this.workTime = workTime;
        this.restTime = restTime;
        this.numReps = numReps;
        this.recoveryTime = recoveryTime;
    }

    //region GETTER methods
    public long getWorkoutSetId() {
        return workoutSetId;
    }

    public long getWorkoutId() {
        return workoutId;
    }

    public long getExerciseTypeId() {
        return exerciseTypeId;
    }

    public int getWorkTime() {
        return workTime;
    }

    public int getRestTime() {
        return restTime;
    }

    public int getNumReps() {
        return numReps;
    }

    public int getRecoveryTime() {
        return recoveryTime;
    }
    //endregion

    //region SETTER methods
    public void setWorkoutSetId(long workoutSetId) {
        this.workoutSetId = workoutSetId;
    }

    public void setWorkoutId(long workoutId) {
        this.workoutId = workoutId;
    }

    public void setExerciseTypeId(long exerciseTypeId) {
        this.exerciseTypeId = exerciseTypeId;
    }

    void setWorkTime(int workTime) {
        this.workTime = workTime;
    }

    void setRestTime(int restTime) {
        this.restTime = restTime;
    }

    void setNumReps(int numReps) {
        this.numReps = numReps;
    }

    void setRecoveryTime(int recoveryTime) {
        this.recoveryTime = recoveryTime;
    }
    //endregion
}

File: app/src/main/java/com/softwareoverflow/hiit_trainer/data/entity/WorkoutSetEntity.kt
package com.softwareoverflow.hiit_trainer.data.entity

import androidx.room.Entity
import androidx.room.ForeignKey

@Entity(
    tableName = "WorkoutSet",
    primaryKeys = ["workoutId", "orderInWorkout"],
    foreignKeys = [
        ForeignKey(
            entity = WorkoutEntity::class,
            parentColumns = ["id"],
            childColumns = ["workoutId"],
            onDelete = ForeignKey.CASCADE,
            onUpdate = ForeignKey.CASCADE
        ),
        ForeignKey(
            entity = ExerciseTypeEntity::class,
            parentColumns = ["id"],
            childColumns = ["exerciseTypeId"],
            onDelete = ForeignKey.CASCADE,
            onUpdate = ForeignKey.CASCADE
        )
    ]
)
class WorkoutSetEntity(
    // The composite primary key is comprised on workoutId and orderInWorkout as no 2 entities can be related to the same workout and have the same position in said workout
    var workoutId: Long,
    var orderInWorkout: Int,

    var exerciseTypeId: Long? = null, // TODO think about the case of deleting exercise types... That probably
    var workTime: Int,
    var restTime: Int,
    var numReps: Int,
    var recoverTime: Int
)